python/ray/data/dataset.py

```diff
@@ -229,6 +229,9 @@ def __init__(
         self._current_executor: Optional["Executor"] = None
         self._write_ds = None
 
+        # Handle sort on empty blocks and still carry the correct batch_format
+        self._batch_format = "default"
+
```

Comment:
Ideally, we would not set this as a Dataset-level attribute. Since the `Dataset` object is meant to represent a fixed set of data, and this attribute should not be modifiable. 

The other alternative I was thinking of is to add a Logical/Physical optimization rule to check for a compatible batch type during logical plan optimization, and directly set `Sort` op's `batch_format` depending on the upstream operators' output batch types. For example, in the added test case, we have 

`FromItems->MapGroups->Sort`

as the LogicalPlan. Then we can add a `LogicalPlan` which traverses up the DAG, finds the compatible `batch_format` of the upstream op (`MapGroups`'s batch_format in this case), and sets the batch format for `Sort` op. We can also do the same for `Shuffle` and `Aggregate` ops in the same rule.  See [ReorderRandomizeBlocksRule](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/logical/rules/randomize_blocks.py) for an example of a logical optimization rule, and [`optimizers.py`](https://github.com/ray-project/ray/blob/master/python/ray/data/_internal/logical/optimizers.py#L22C5-L22C31) on how to add the rule into the plan optimizer.

@xingyu-long let me know if this makes sense, and if you have any other questions. Thanks!

Reply:
@scottjlee Thanks for the suggestion! It makes sense with more general solution. I will spend some time and look into it with above sample code.

Reply:
Hi @scottjlee, I just implemented the way as you suggested and it's much cleaner and general! but I added some TODOs just for now since I'd like to give you first look for what I did and see if it's same idea as you suggested. After  confirmed with you, I will fill all TODOs I left and tests as well.  Could you take a look when you have time? Thanks!